src/controllers/restful/PaymentController.ts

```diff
@@ -0,0 +1,230 @@
+import { PaymentsCheckRequest } from "../../types/controllers/PaymentsCheckRequest";
+import { PaymentsCheckResponse } from "../../types/controllers/PaymentsCheckResponse";
+import { CodiceContestoPagamento } from "../../types/PagoPaTypes";
+import * as PaymentsConverter from "../../utils/PaymentsConverter";
+
+/**
+ * This controller is invoked by BackendApp
+ * to retrieve information about a qrcode (payment).
+ * It asks PagoPa for payment information using VerificaRPT service.
+ * @param {express.Request} req - The RESTful request
+ * @param {PagoPaConfig} pagoPaConfig - Configuration about PagoPa WS to contact
+ * @return {Promise<PaymentCtrlResponse<PaymentsActivationResponse>>} The response content to send to applicant
+ */
+export function checkPaymentToPagoPa(
+  pagoPaConfig: PagoPaConfig,
+  paymentVerificaRPTPagoPaClient: pagoPaSoapClient.PagamentiTelematiciPspNodoAsyncClient
+): (
+  req: express.Request
+) => Promise<
+  | IResponseErrorValidation
+  | IResponseSuccessJson<PaymentsCheckResponse>
+  | IResponseErrorGeneric
+> {
+  return async req => {
+    // Validate qrcode data provided by BackendApp
+    const errorOrPaymentsCheckRequest = PaymentsCheckRequest.decode(req.params);
+    if (isLeft(errorOrPaymentsCheckRequest)) {
+      const error = errorOrPaymentsCheckRequest.value;
+      return ResponseErrorFromValidationErrors(PaymentsCheckRequest)(error);
+    }
+    const paymentsCheckRequest = errorOrPaymentsCheckRequest.value;
+
+    // Generate a Session Token called CodiceContestoPagamento
+    // to follow a stream of requests with PagoPa.
+    // It will be generated here after the first interaction
+    // started by BackendApp (checkPaymentToPagoPa)
+    // For the next messages, BackendApp will provide the same codiceContestoPagamento
+    const codiceContestoPagamento = generateCodiceContestoPagamento();
+
+    // Convert the input provided by BackendApp (RESTful request) to a PagoPa request (SOAP request).
+    // Some static information will be obtained by pagoPaConfig, to identify this client.
+    const errorOrPaymentCheckRequestPagoPa = PaymentsConverter.getPaymentsCheckRequestPagoPa(
+      pagoPaConfig,
+      paymentsCheckRequest,
+      codiceContestoPagamento
+    );
+    if (isLeft(errorOrPaymentCheckRequestPagoPa)) {
+      const error = errorOrPaymentCheckRequestPagoPa.value;
+      return ResponseErrorValidation("Invalid PagoPa Request", error.message);
+    }
+    const paymentCheckRequestPagoPa = errorOrPaymentCheckRequestPagoPa.value;
+
+    // Send the SOAP request to PagoPa (VerificaRPT message)
+    const errorOrPaymentCheckPagoPaResponse = await PaymentsService.sendPaymentCheckRequestToPagoPa(
+      paymentCheckRequestPagoPa,
+      paymentVerificaRPTPagoPaClient
+    );
+    if (isLeft(errorOrPaymentCheckPagoPaResponse)) {
+      const error = errorOrPaymentCheckPagoPaResponse.value;
+      return ResponseErrorGeneric(
```

Comment:
Since it's a common (500) error I'd rather use here 
```ts
return ResponseErrorInternal(`PagoPa Server communication error: ${error.message}`)
```